### DISCUSSION TITLE
질문

### DISCUSSION THREAD

**Discussion by Hsg2311** [MyeongWoonJang/cpp-study#90] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/75879925/200611854-3a2587d5-964e-4a80-8488-7144e36c1d3a.png)
a > b 일 때, true가 return되면 a와 b를 swap한다는 의미인지, 아니면 그냥 넘어간다는 의미인지 모르겠습니다.

**Comment by Woon-2** [comment, marked as answer]
`std::sort`는 `introsort(큰 분할 단위에서 quicksort, 작은 분할 단위에서 insertionsort, 일정 재귀 깊이 이상에서 heapsort)` 알고리즘을 사용하기 때문에 `predicate(조건자)`가 `true`를 반환하다고 바로 `swap`시키진 않습니다.

일반적으로 `STL` 알고리즘 함수들에 자주 등장하는 매개변수인 `predicate`는 지정하지 않을 경우 `std::less` 객체를 이용하게 되어있습니다.   
`std::less`는 함수 객체입니다.   
간략화하면, 다음과 같은 생김새를 지녔습니다.

```cpp
<template class T>
class less
{
public:
    bool operator()(const T& lhs, const T& rhs)
    {
        return lhs < rhs;
    }
};
```

물론 실제 컴파일러의 구현은 훨씬 더 타입을 정교하게 다룰 것입니다.   
그러나 `std::less`의 골자는 기본적으로 위와 같습니다.

따라서, 어떠한 `predicate`를 람다 표현식이나 함수 객체 등으로 직접 만들어서 전달할 경우에는,   
`predicate`의 기본값이 `std::less`로 상정되어있다는 것을 떠올리시면 됩니다.

`std::sort`의 기본 동작은 오름차순 정렬이죠.   
질문자분께서 만드신 람다 표현식은 `std::less`와 비교했을 때 왼쪽 값이 오른쪽 값보다 커야 `true`를 반환하므로, 내림차순을 유도하게 될 것입니다.

람다 표현식이 들어간 자리에 `std::greater<>{}`를 써보시면 정확히 같은 결과를 얻을 수 있습니다.   
(`operator<`가 정의되어 있다면 말이죠.)

`std::sort`의 경우엔, `predicate`의 의미를 이렇게 해석할 수 있겠네요.   
"왼쪽 값과 오른쪽 값 중 어느 값이 더 커야 정렬된 상태로 판별할 것인가"

질문자분이 만드신 람다 표현식은 왼쪽 값이 오른쪽 값보다 클 때 `true`를 반환하니,   
왼쪽 값이 오른쪽 값보다 클 때를 정렬된 상태로 판별하겠다는 의지를 보이신 겁니다.

그 결과는 내림차순 정렬이 되겠지요.

**Reply by Woon-2** [reply]
함수 객체라는 것이 생소할 수 있으니,   
`std::less`가 왜 `predicate`의 기본값인지 설명할 겸 좀 깊은 내용을 얘기해보려 합니다.

`std::less`가 `STL` 알고리즘들에 필요한 `predicate`의 기본값인 것은,    
`std::less`만으로 다른 모든 비교 연산을 만들어낼 수 있기 때문입니다.

다음 코드는 `my_unique_ptr`가 참여하는 비교 연산자 오버로딩들입니다.

https://github.com/Woon-2/cpp-study/blob/3de0db2eb8d61f4d558b5ff11458c014e4551b4a/solutions-sprint1-write-cpp-classes/my_unique_ptr.inl#L108-L130

보시다시피 `operator<`만 `std::less`를 써서 직접적으로 구현하고,   
`operator>`과 `operator<=`, `operator>=`는 `operator<`를 재사용하여 구현합니다.   
매우 짧은 함수들이기 때문에 이들은 인라인될 가능성이 높고,   
실제 컴파일된 코드에서 연산자를 오버로드한 함수의 호출은 모두 사라집니다.   

```cpp
my_unique_ptr<Foo> lhs;    // Foo라는 어떤 클래스의 선언과 정의가 이미 마련되어 있다고 가정합니다.
my_unique_ptr<Foo> rhs;

// 여기서 어떤 작업이 한창 진행되었습니다.

if (lhs < rhs)    // lhs < rhs는 operator<(lhs, rhs)와 같습니다.
    do_something();    // 무엇인가 합니다.

```

대충 이런 흐름의 코드가 있다고 가정해봅시다.   
`lhs < rhs`는 `operator<(lhs,rhs)`와 같고, 불러야 할 오버로딩이자 템플릿 특수화가 어떤 것인가 찾아보니    
`operator<(const my_unique_ptr<Foo>&, const my_unique_ptr<Foo>&)`입니다.   

그리고 이 함수의 정의를 보니 인라인하기 매우 좋아보입니다.   
컴파일러가 이 함수를 인라인하고 나면 코드는 다음과 같이 바뀐 셈입니다.

```cpp
if (std::less<std::common_type_t<Foo*, Foo*>>{}(lhs.get(), rhs.get()))
    do_something();
```

무언가 길어서 엄청나게 많은 일을 할 것 같지만 실상은 아닙니다.   
`std::common_type`는 상속 관계에 있을지도 모르는 타입들에 대해 "가장 일반적인(상속의 `Base`가 되는)" 타입을 추출해주는 `helper struct`이고,   
그 추출 결과는 `std::common_type<T...>::type`과 같이 꺼내서 써야 하는데,   
타이핑을 줄일 수 있도록 `using` 키워드를 사용한 것이 `std::common_type_t<T...>`입니다.

이렇게 생겼어요.

```cpp
template<class ... T>
using common_type_t = typename common_type<T...>::type
```

코드 상에 적힌 모든 타입은 적어도 `compile time`에는 확정이 끝납니다.   
따라서 `run time`에 아무런 `overhead`가 없어요.

사실상 다음과 비용 차이가 없다는 말입니다.

```cpp
if (std::less<Foo*>{}(lhs.get(), rhs.get())
    do_something();
```

여기서 끝이 아닙니다.

함수 객체마저도 인라인될 수 있습니다.   
함수 객체를 사용하는 이유 중에 하나이지요.   

함수 포인터로는 호출할 함수가 뭔지 알 수 없어 인라인할 수 없지만,   
함수 객체는 그 클래스 정의에 어떤 함수를 호출할지 나와있습니다.   
(`operator()`의 오버로딩이 바로 그것이죠.)

그 함수가 인라인하기 적합한 함수라면, 역시 인라인됩니다.

```cpp
if (lhs.get() < rhs.get())
    do_something();
```

그럼 이제 끝났을까요?   

`get()`의 인라인 여부도 살펴봐야죠.   

`Foo* const my_unique_ptr<Foo>::get() const`은 `Foo* const my_unique_ptr<Foo>::operator->() const`를 호출하고   
`Foo* const my_unique_ptr<Foo>::operator->() const`는 멤버 변수인 `ptr`을 반환합니다.

https://github.com/Woon-2/cpp-study/blob/3de0db2eb8d61f4d558b5ff11458c014e4551b4a/solutions-sprint1-write-cpp-classes/my_unique_ptr.inl#L52-L56

https://github.com/Woon-2/cpp-study/blob/3de0db2eb8d61f4d558b5ff11458c014e4551b4a/solutions-sprint1-write-cpp-classes/my_unique_ptr.inl#L40-L44

역시 컴파일러가 인라인에 입맛을 다실 만한 짧은 함수들이지요.   
다음의 인라인 단계를 거칩니다.

먼저, `get()`을 인라인합니다.

```cpp
if (lhs.operator->() < rhs.operator->())
    do_something();
```

그리고, `operator->()`을 인라인합니다.

```cpp
if (lhs.ptr < rhs.ptr)
    do_something();
```

다른 연산자들도 마찬가집니다.

```cpp
lhs > rhs       // rhs.ptr < lhs.ptr으로 인라인될 것입니다.
lhs <= rhs    // !(rhs.ptr < lhs.ptr)으로 인라인될 것입니다.
lhs >= rhs    // !(lhs.ptr < rhs.ptr)으로 인라인될 것입니다.
```

1. `operator<(lhs, rhs)`
2. `std::less<Foo*>{}(lhs.get(), rhs.get())`
3. `lhs.get()`
4. `rhs.get()`
5. `lhs.operator->()`
6. `rhs.operator->()`

함수 호출이 무려 6개나 생략되었습니다!   
6개의 스택 프레임이 `push`되고 `pop`되는 과정이 마치 처음부터 없었던 것처럼 말끔히 사라져버렸습니다!

`operator>`, `operator<=`, `operator>=`는 내부적으로 `operator<`를 이용하기 때문에   
이들의 경우에는 함수 호출이 7개가 생략된 것이죠!

이것이 `C++`가 가진 힘입니다.   
모든 함수가 가상 함수로 동작하는 J모 언어, P모 언어에서는 절대로 이런 것을 할 수 없습니다.

가상 함수는 `run time`에 어떤 함수를 호출할지 결정되는데,   
(`vtbl`에 담긴 함수 포인터를 이용합니다, 함수 포인터를 통한 호출은 그 함수의 본문을 알 수 없어 인라인할 수 없죠.)   
어떻게 `compile time`에 호출할 함수를 인라인할 수 있나요, 불가능합니다.
